src/Effect/Exec.hs

```diff
@@ -129,17 +132,29 @@ instance (Algebra sig m, MonadIO m) => Algebra (Exec :+: sig) (ExecIOC m) where
 
       let cmdName' = T.unpack $ cmdName cmd
           cmdArgs' = map T.unpack $ cmdArgs cmd
-      (exitcode, stdout, stderr) <- readProcess (setWorkingDir (fromAbsDir absolute) (proc cmdName' cmdArgs'))
-
-      let failure = CmdFailure (cmdName cmd) (cmdArgs cmd) (fromAbsDir absolute) exitcode stdout stderr
-
-      let result = case (exitcode, cmdAllowErr cmd) of
-            (ExitSuccess, _) -> Right stdout
-            (_, Never) -> Left failure
-            (_, NonEmptyStdout) ->
-              if BL.null stdout
-                then Left failure
-                else Right stdout
-            (_, Always) -> Right stdout
+
+          mkFailure :: ExitCode -> Stdout -> Stderr -> CmdFailure
+          mkFailure = CmdFailure (cmdName cmd) (cmdArgs cmd) (fromAbsDir absolute)
+
+          ioExceptionToCmdFailure :: IOException -> CmdFailure
+          ioExceptionToCmdFailure = mkFailure (ExitFailure 1) "" . fromString . show
+
+      processResult <- try $ readProcess (setWorkingDir (fromAbsDir absolute) (proc cmdName' cmdArgs'))
+
+      -- apply business logic for considering whether exitcode + stderr constitutes a "failure"
+      let mangleResult :: (ExitCode, Stdout, Stderr) -> Either CmdFailure Stdout
+          mangleResult (exitcode, stdout, stderr) =
+            case (exitcode, cmdAllowErr cmd) of
+                  (ExitSuccess, _) -> Right stdout
+                  (_, Never) -> Left $ mkFailure exitcode stdout stderr
+                  (_, NonEmptyStdout) ->
+                    if BL.null stdout
+                      then Left $ mkFailure exitcode stdout stderr
+                      else Right stdout
+                  (_, Always) -> Right stdout
+
+
+      let result :: Either CmdFailure Stdout
+          result = first ioExceptionToCmdFailure processResult >>= mangleResult
```

Comment:
`first` comes from `Bifunctor`, which is like `Functor` but allows mapping over two positional arguments:
```haskell
class Bifunctor (p :: Type -> Type -> Type) where
  bimap :: (a -> b) -> (c -> d) -> p a c -> p b d

  first :: (a -> b) -> p a c -> p b c
  first f = bimap f id

  second :: (b -> c) -> p a b -> p a c
  second g = bimap id g
```

in this case, `first` is monomorphized to:
```haskell
first ::
  (IOException -> CmdFailure) ->
  Either IOException Stdout ->
  Either CmdFailure Stdout
```